tests/ops/op_math/test_controlled_ops.py

```diff
@@ -453,176 +457,492 @@ def test_unitary_check(self):
-    (qml.CZ(wires=(0, 1)), Wires(0)),
-]
+        assert np.allclose(op.compute_eigvals(*params), expected_eigvals, atol=tol, rtol=0)
+        assert np.allclose(op(*params, wires=[0, 1]).eigvals(), expected_eigvals, atol=tol, rtol=0)
 
+    @pytest.mark.tf
+    @pytest.mark.parametrize("op, params, expected_eigvals", EXPECTED_EIGVALS)
+    def test_eigvals_tf(self, tol, op, params, expected_eigvals):
+        """Tests that the correct eigenvalues are returned when using TensorFlow"""
 
-@pytest.mark.parametrize("op, control_wires", control_data)
-def test_control_wires(op, control_wires):
-    """Test ``control_wires`` attribute for non-parametrized operations."""
+        import tensorflow as tf
 
-    assert op.control_wires == control_wires
+        variables = [tf.Variable(param) for param in params]
 
+        assert np.allclose(op.compute_eigvals(*variables), expected_eigvals, atol=tol, rtol=0)
+        assert np.allclose(
+            op(*variables, wires=[0, 1]).eigvals(), expected_eigvals, atol=tol, rtol=0
+        )
 
-involution_ops = [  # ops who are their own inverses
-    qml.CY((0, 1)),
-    qml.CZ(wires=(0, 1)),
-]
 
+def _verify_decompositions(ops, classes, params, wires):
+    """Verifies that the correct decomposition classes are returned"""
+    for op, c, p, w in zip(ops, classes, params, wires):
+        assert isinstance(op, c)
+        assert qml.math.allclose(op.parameters, p)
+        assert op.wires == w
 
-@pytest.mark.parametrize("op", involution_ops)
-def test_adjoint_method(op):
-    """Tests the adjoint method for operations that are their own adjoint."""
-    adj_op = copy.copy(op)
-    for _ in range(4):
-        adj_op = adj_op.adjoint()
 
-        assert qml.equal(adj_op, op)
+def _dot_broadcasted(a, b):
+    return np.einsum("...ij,...jk->...ik", a, b)
 
 
-@pytest.mark.parametrize("op_cls, _", NON_PARAMETRIZED_OPERATIONS)
-def test_map_wires(op_cls, _):
-    """Test that we can get and set private wires in all operations."""
+def _multi_dot_broadcasted(matrices):
+    return functools.reduce(_dot_broadcasted, matrices)
 
-    op = op_cls(wires=[0, 1])
-    assert op.wires == Wires((0, 1))
 
-    op = op.map_wires(wire_map={0: "a", 1: "b"})
-    assert op.base.wires == Wires(("b"))
-    assert op.control_wires == Wires(("a"))
+class TestDecompositions:
+    @pytest.mark.parametrize("phi", [0.432, np.array([0.1, 2.1])])
+    def test_CRX(self, phi):
+        """Test the decomposition for CRX."""
+
+        ops1 = qml.CRX.compute_decomposition(phi, wires=[0, 1])
+        ops2 = qml.CRX(phi, wires=(0, 1)).decomposition()
+
+        classes = [qml.RZ, qml.RY, qml.CNOT, qml.RY, qml.CNOT, qml.RZ]
+        params = [[np.pi / 2], [phi / 2], [], [-phi / 2], [], [-np.pi / 2]]
+        wires = [Wires(1), Wires(1), Wires((0, 1)), Wires(1), Wires((0, 1)), Wires(1)]
+
+        _verify_decompositions(ops1, classes, params, wires)
+        _verify_decompositions(ops2, classes, params, wires)
```

Comment:
```suggestion
        expected = [qml.RZ(np.p/2, 1), qml.RY(phi/2, 1), qml.CNOT((0,1)), qml.RY(-phi/2, 1), qml.CNOT((0,1)), qml.RZ(-np.pi/2, 1)
        assert ops1 == expected
        assert op2 == expected
```

We can just do this now. Comparison with `==` now uses `qml.equal`.

Reply:
Done